larpandoracontent/LArMonitoring/VisualParticleMonitoringAlgorithm.cc

```diff
@@ -0,0 +1,428 @@
+
+void VisualParticleMonitoringAlgorithm::VisualizeIndependentMC(const LArMCParticleHelper::MCContributionMap &mcMap)
+{
+    std::map<int, const std::string> keys = {{13, "mu"}, {11, "e"}, {22, "gamma"}, {321, "kaon"}, {211, "pi"}, {2212, "p"}};
+    std::map<int, Color> colors = {{0, RED}, {1, BLACK}, {2, BLUE}, {3, CYAN}, {4, MAGENTA}, {5, GREEN}, {6, ORANGE}, {7, GRAY}};
+    MCParticleList linearisedMC;
+    if (mcMap.empty())
+        return;
+
+    PANDORA_MONITORING_API(SetEveDisplayParameters(this->GetPandora(), true, DETECTOR_VIEW_XZ, -1.f, 1.f, 1.f));
+    LArMCParticleHelper::GetBreadthFirstHierarchyRepresentation(mcMap.begin()->first, linearisedMC);
+
+    int colorIdx{0}, mcIdx{0};
+    for (const MCParticle *pMC : linearisedMC)
+    {
+        const auto iter{mcMap.find(pMC)};
+        if (iter == mcMap.end())
+            continue;
+        std::string key("other");
+        try
+        {
+            const int pdg{std::abs(pMC->GetParticleId())};
+            if (keys.find(pdg) != keys.end())
+                key = keys[pdg];
+        }
+        catch (const StatusCodeException&)
+        {
+            key = "unknown";
+        }
+
+        CaloHitList uHits, vHits, wHits;
+        for (const CaloHit *pCaloHit : iter->second)
+        {
+            const HitType view{pCaloHit->GetHitType()};
+            if (view == HitType::TPC_VIEW_U)
+                uHits.emplace_back(pCaloHit);
+            else if (view == HitType::TPC_VIEW_V)
+                vHits.emplace_back(pCaloHit);
+            else
+                wHits.emplace_back(pCaloHit);
+        }
+        std::string suffix{std::to_string(mcIdx) + "_" + key};
+        if (!uHits.empty())
+            PANDORA_MONITORING_API(VisualizeCaloHits(this->GetPandora(), &uHits, "u_" + suffix, colors[colorIdx]));
+        if (!vHits.empty())
+            PANDORA_MONITORING_API(VisualizeCaloHits(this->GetPandora(), &vHits, "v_" + suffix, colors[colorIdx]));
+        if (!wHits.empty())
+            PANDORA_MONITORING_API(VisualizeCaloHits(this->GetPandora(), &wHits, "w_" + suffix, colors[colorIdx]));
+        colorIdx = colorIdx >= colors.size() ? 0 : colorIdx + 1;
+        ++mcIdx;
+    }
+
+    PANDORA_MONITORING_API(ViewEvent(this->GetPandora()));
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void VisualParticleMonitoringAlgorithm::VisualizeMCByPdgCode(const LArMCParticleHelper::MCContributionMap &mcMap)
+{
+    std::map<int, const std::string> keys = {{13, "mu"}, {11, "e"}, {22, "gamma"}, {321, "kaon"}, {211, "pi"}, {2212, "p"}};
+    std::map<std::string, Color> colors = {{"mu", MAGENTA}, {"e", RED}, {"gamma", ORANGE}, {"kaon", BLACK}, {"pi", GREEN}, {"p", BLUE},
+        {"other", GRAY}};
+
+    std::map<std::string, CaloHitList> uHits, vHits, wHits;
+    for (const auto [ key, value ] : keys)
+    {
+        (void)key;  // GCC 7 support, 8+ doesn't need this
```

Comment:
Good to note, and to look out for in our different build use cases.

Reply:
Yes, this scenario has been seen before and our CI is GCC 7, so this does get picked up.